dataflow/dofn.py

```diff
@@ -0,0 +1,162 @@
+    return int(datetime.strptime(row['time'], "%m/%d/%Y %H:%M:%S")
+               .replace(tzinfo=ZoneInfo("America/Los_Angeles")).timestamp())
+
+
+class WeatherTransformDoFn(DoFn):
+    __city_to_segments: Dict[str, Set[int]] = dict()
+
+    def __init__(self, segments: List[Segment]):
+        super().__init__()
+        for segment in segments:
+            if segment['city'] not in self.__city_to_segments:
+                self.__city_to_segments[segment['city']] = set()
+            self.__city_to_segments.get(segment['city']).add(segment['id'])
+
+    def process(self, row, *args, **kwargs):
+        ts = get_weather_timestamp(row)
+        for id in self.__city_to_segments.get(row['name'], set()):
+            yield TimestampedValue((id, row), ts)
+
+
+class BayArea511EventTransformDoFn(DoFn):
+    __segment_to_coord: Dict[int, Tuple[float, float]] = dict()
+    MAXIMUM_DISTANCE_MILES = 10
+
+    def __init__(self, segments: List[Segment]):
+        super().__init__()
+        for segment in segments:
+            self.__segment_to_coord[segment['id']] = segment['representative_point']
+
+    def process(self, row, *args, **kwargs):
+        ts = get_bay_area_511_event_timestamp(row)
+        for id, coord in self.__segment_to_coord.items():
+            distance = geopy.distance.geodesic(reversed(row['geography_point']['coordinates']),
+                                               coord).miles  # PeMS coord are in [lat, lon] but we require [lon, lat]
+            if distance <= self.MAXIMUM_DISTANCE_MILES:
+                yield TimestampedValue((id, row), ts)
+
+
+class PeMSTransformDoFn(DoFn):
+    __station_to_segment: Dict[int, Set[int]] = dict()
+
+    def __init__(self, segments: List[Segment]):
+        super().__init__()
+        for segment in segments:
+            for station_id in segment['station_ids']:
+                if int(station_id) not in self.__station_to_segment:
+                    self.__station_to_segment[int(station_id)] = set()
+                self.__station_to_segment.get(int(station_id)).add(segment['id'])
+
+    def process(self, row, *args, **kwargs):
+        ts = get_pems_timestamp(row)
+        for id in self.__station_to_segment.get(row['station_id'], set()):
+            yield TimestampedValue((id, row), ts)
+
+
+class SegmentFeatureTransformDoFn(DoFn):
+    __segments: List[Segment] = []
+    __metadata_version: int
+
+    def __init__(self, segments: List[Segment], metadata_version: int):
+        super().__init__()
+        self.__segments = segments
+        self.__metadata_version = metadata_version
+
+    def process(self, element, window=DoFn.WindowParam):
+        segment_id, data = element
+        t = window.end
+        features = self.get_event_features(data['bay_area_511_event'], segment_id, t) + \
+                   self.get_pems_feature(data['pems'], segment_id) + \
+                   self.get_weather_features(data['weather']) + \
+                   self.get_time_features(t)
```

Comment:
For the correct model prediction, the order should be 
                   `self.get_pems_feature(data['pems'], segment_id) + \
                   self.get_event_features(data['bay_area_511_event'], segment_id, t) + \
                   self.get_weather_features(data['weather']) + \
                   self.get_time_features(t)`

Reply:
Upon further review, it appears that the order you mentioned is incorrect. The correct sequence should be as I have implemented in my code, which aligns with the [notebook you shared](https://cloud-city-cal.slack.com/archives/C05SPDA2R9U/p1705807185488819?thread_ts=1705719240.149279&cid=C05SPDA2R9U). You can examine the order of the coefficients with `model.feature_names_in_`.

![image](https://github.com/BayAreaCloudCity/gcp/assets/12138874/4180b822-8566-4ffd-b3e2-88db11dab782)

The previous issue of erroneous output was due to the model being trained with `fit_intercept=True`. However, the JSON file for coefficients did not include the intercept values, leading to the predictions always being off by 6 million. To fix this, because of one-hot encoding, including the intercept should NOT happen, as it would lead to co-linearity, causing the coefficients to be nonsensical. After setting `fit_intercept` to `False`, the coefficients now make logical sense:

```
5.645033060419449,-5.66662515522065e-12,5.908583704673055,2.842170943040401e-14,1.0658141036401503e-14,5.808699640348686,-0.08034505132084768,0.8266674774762848,-3.1086244689504383e-15,2.8767626374910833,1.981673688771794,-1.3322676295501878e-15,2.322872531483047,2.3314683517128287e-15,2.8584559561432483,0.0,2.216266716085547,2.220446049250313e-16,0.0,7.216449660063518e-16,-1.6930901125533637e-15,2.4811151748374694,2.625169700629091,3.395807159956021e-05,2.5943419815289093,2.3349915669941455,2.311721907340515,2.232322193818984,2.317392432717097,2.679897642213585,2.8916486808280184,1.6572623474291577,1.8224089393866425,1.643516179489951,1.6917423934148337,1.5571718345409225,1.058924872877516,0.42077274437130796,0.21821648398809845,1.284655986643573,1.2402773176043926,0.4235076765011194,0.47164656538114347,-0.01854937290502301,-1.9349889599724788,-2.8175143552355832,-2.062703937602055,-2.8039581448046746,-1.457768430743203,3.8691125471542733,3.1220874839734045,2.4318593866563525,1.7267153869320477,1.870461301562459,1.947460158797133
```